python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,505 @@
+
+        self.nsamples = 2 * self.M + 2 ** 12 if nsamples is None else nsamples
+
+        self.max_samples = 2 ** 30
+
+        # restricting maximum number of samples for memory and performance
+        # value being checked, right now based on mainline SHAP package
+        self.max_samples = 2 ** 30
+        if self.M <= 30:
+            self.max_samples = 2 ** self.M - 2
+            if self.nsamples > self.max_samples:
+                self.nsamples = self.max_samples
+
+        if isinstance(data, pd_df) or isinstance(data, cu_df):
+            self.feature_names = data.columns.to_list()
+        else:
+            self.feature_names = [None for _ in range(len(data))]
+
+        cur_nsamples = 0
+        self.nsamples_exact = 0
+        r = 0
+
+        # we check how many subsets of the powerset of self.M we can fit
+        # in self.nsamples
+        while cur_nsamples <= self.nsamples:
+            r += 1
+            self.nsamples_exact = cur_nsamples
+            cur_nsamples += int(binomCoef(self.M, r))
+
+        # see if we need to have randomly sampled entries in our mask
+        # and combinations matrices
+        self.nsamples_random = \
+            self.nsamples - self.nsamples_exact if r < self.M else 0
+
+        # we save r so we can generate random samples later
+        self.randind = r
+
+        # using numpy powerset and calculations for initial version
+        # cost is incurred only once, and generally we only generate
+        # very few samples if M is big.
+        mat, weight = powerset(self.M, r - 1, self.nsamples_exact,
+                               dtype=self.dtype)
+
+        self.mask = cp.zeros((self.nsamples, self.M), dtype=np.float32)
+        self.mask[:self.nsamples_exact] = cp.array(mat)
+
+        self.weights = cp.ones(self.nsamples, dtype=self.dtype)
+        self.weights[:self.nsamples_exact] = cp.array(weight)
+
+        self.synth_data = None
+
+        self.expected_value = self.link_fn(cp.mean(model(self.background)))
+
+        self.random_state = random_state
+
+    def explain(self,
+                X,
+                nsamples=None,
+                l1_reg='auto'):
+        shap_values = cp.zeros((1, self.M), dtype=self.dtype)
+
+        if X.ndim == 1:
+            X = X.reshape((1, self.M))
+
+        # allocating combinations array once for multiple explanations
+        if self.synth_data is None:
+            self.synth_data = cp.zeros(
+                shape=(self.N * self.nsamples, self.M),
+                dtype=np.float32,
+                order='C'
+            )
+
+        idx = 0
+        for x in X:
+            shap_values[idx] = self._explain_single_observation(
+                x.reshape(1, self.M), l1_reg
+            )
+
+        if isinstance(X, np.ndarray):
+            out_type = 'numpy'
+        else:
+            out_type = 'cupy'
+        return input_to_cuml_array(shap_values)[0].to_output(out_type)
```

Comment:
```suggestion
        return shap_values[0]
```
If you add the decorator, this will perform the conversion for you. (FYI, this will return other types than cupy or numpy. If you want to to work identically, let me know. Easy fix.)